### PR TITLE
environment_check: check per node packages

### PR DESCRIPTION
**Pull request by derekbit**
Longhorn 3550

Signed-off-by: Derek Su <derek.su@suse.com>

### CODE REVIEW

File: deploy/scripts/environment_check.sh [hunk 1]
```diff
@@ -100,10 +114,54 @@ validate_ds() {
...
+  local packages=($@)
+  local pods=$(kubectl get pods -o name | grep longhorn-environment-check)
+
+  local allFound=true
+
+  for ((i=0; i<${#packages[@]}; i++)); do
+    local package=${packages[$i]}
+
+    for pod in ${pods}; do
+      installed=`kubectl exec -it $pod -- nsenter --mount=/proc/1/ns/mnt -- bash -c "dpkg --get-selections ${package}" | grep -c 'install'`
```
Comment (innobead) [inline deploy/scripts/environment_check.sh:127]: This should also work for all distros we support in this script.

Comment (derekbit) [reply deploy/scripts/environment_check.sh:127]: Oops, sorry, forgot to handle `dpkg ...`

File: deploy/scripts/environment_check.sh [hunk 2]
```diff
@@ -65,45 +144,105 @@ wait_ds_ready() {
...
+  CHECK_CMD="rpm -q"
+  packages=(nfs-utils iscsi-initiator-utils)
+  ;;
+*"suse"* )
+  CHECK_CMD="zypper se -i"
+  packages=(nfs-client open-iscsi)
+  ;;
+*)
+  CHECK_CMD=""
+  packages=()
```
Comment (innobead) [inline deploy/scripts/environment_check.sh:230]: If not supported OS, suggest to error out and exit.

**Review by innobead — COMMENTED** [review]
LGTM, but it seems we don't check if multipass is running, do we need?

File: deploy/scripts/environment_check.sh (hunk 2 above)
Comment (derekbit) [reply deploy/scripts/environment_check.sh:230]: Okay.

**Comment by derekbit** [comment]
> LGTM, but it seems we don't check if multipass is running, do we need?

I'm thinking how to check multipath on the node. Maybe just check the `multipath.conf` has correct blacklist?
https://longhorn.io/kb/troubleshooting-volume-with-multipath/

cc. @PhanLe1010

**Comment by innobead** [comment]
> > LGTM, but it seems we don't check if multipass is running, do we need?
> 
> I'm thinking how to check multipath on the node. Maybe just check the `multipath.conf` has a correct blacklist? [longhorn.io/kb/troubleshooting-volume-with-multipath](https://longhorn.io/kb/troubleshooting-volume-with-multipath/)
> 
> cc. @PhanLe1010

or just check if the multipass is running, show a warning to inform users there will be probably some problem about volumes automatically mounted? I don't think we need to specially identify the problem. WDYT?

File: deploy/scripts/environment_check.sh (hunk 2 above)
Comment (derekbit) [reply deploy/scripts/environment_check.sh:230]: > > > LGTM, but it seems we don't check if multipass is running, do we need?
> > 
> > 
> > I'm thinking how to check multipath on the node. Maybe just check the `multipath.conf` has a correct blacklist? [longhorn.io/kb/troubleshooting-volume-with-multipath](https://longhorn.io/kb/troubleshooting-volume-with-multipath/)
> > cc. @PhanLe1010
> 
> or just check if the multipass is running, show a warning to inform users there will be probably some problem about volumes automatically mounted? I don't think we need to specially identify the problem. WDYT?

Good idea!

**Comment by PhanLe1010** [comment]
Yeah, agree with @innobead that we can just check if multipath is running and show a warning message with the link to the GitHub issue

**Comment by PhanLe1010** [comment]
I am wondering if it is possible that the host has the open-iscsi package but the iscsid is not running?

**Comment by derekbit** [comment]
> I am wondering if it is possible that the host has the open-iscsi package but the iscsid is not running?

Yes, I think we need this. I just remember there is an issue that the iscsid was not started after installing the iscsi package from an user.

**Comment by PhanLe1010** [comment]
Thanks @derekbit Can you help to add the check for iscsid?

**Comment by derekbit** [comment]
> Thanks @derekbit Can you help to add the check for iscsid?

Sure thing. I'm working on it.

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, just minor suggestions

**Review by PhanLe1010 — DISMISSED** [review]
LGTM

**Review by innobead — APPROVED** [review]

**Comment by jenting** [comment]
@derekbit 
Looks like we need to update the https://github.com/longhorn/longhorn/blob/master/scripts/environment_check.sh as well.

**Comment by innobead** [comment]
> @derekbit Looks like we need to update the [https://github.com/longhorn/longhorn/blob/master/scripts/environment_check.sh](https://github.com/longhorn/longhorn/blob/master/scripts/environment_check.sh?rgh-link-date=2022-02-10T06%3A33%3A22Z) as well.

YES, but I am planning to keep only one place. Let me handle that.

**Comment by jenting** [comment]
Besides, we could consider improving the doc output https://longhorn.io/docs/1.2.3/deploy/install/#using-the-environment-check-script of the 1.3.0 version.

**Pull request merged**